codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java

```diff
@@ -168,31 +218,93 @@ public void dataWriteAfterGoAwayShouldFail() throws Exception {
-        } finally {
-            data.release();
-        }
+        encoder.writeData(ctx, STREAM_ID, data, 0, true, promise);
+        assertEquals(1, payloadCaptor.getAllValues().size());
+        // Write the DATA frame completely
+        payloadCaptor.getValue().write(Integer.MAX_VALUE);
+        verify(lifecycleManager).closeLocalSide(eq(stream), eq(promise));
+        assertEquals(0, data.refCnt());
+    }
+
+    @Test
+    public void dataLargerThanMaxFrameSizeShouldBeSplit() throws Http2Exception {
+        when(frameSizePolicy.maxFrameSize()).thenReturn(3);
+        final ByteBuf data = dummyData();
+        encoder.writeData(ctx, STREAM_ID, data, 0, true, promise);
+        assertEquals(payloadCaptor.getValue().size(), 8);
+        assertTrue(payloadCaptor.getValue().write(8));
+        // writer was called 3 times
+        assertEquals(3, writtenData.size());
+        assertEquals("abc", writtenData.get(0));
+        assertEquals("def", writtenData.get(1));
+        assertEquals("gh", writtenData.get(2));
+        assertEquals(0, data.refCnt());
+    }
+
+    @Test
+    public void paddingSplitOverFrame() throws Http2Exception {
+        when(frameSizePolicy.maxFrameSize()).thenReturn(5);
+        final ByteBuf data = dummyData();
+        encoder.writeData(ctx, STREAM_ID, data, 5, true, promise);
+        assertEquals(payloadCaptor.getValue().size(), 13);
+        assertTrue(payloadCaptor.getValue().write(13));
+        // writer was called 3 times
+        assertEquals(3, writtenData.size());
+        assertEquals("abcde", writtenData.get(0));
+        assertEquals(0, (int) writtenPadding.get(0));
+        assertEquals("fgh", writtenData.get(1));
+        assertEquals(2, (int) writtenPadding.get(1));
+        assertEquals("", writtenData.get(2));
+        assertEquals(3, (int) writtenPadding.get(2));
+        assertEquals(0, data.refCnt());
+    }
+
+    @Test
+    public void frameShouldSplitPadding() throws Http2Exception {
+        when(frameSizePolicy.maxFrameSize()).thenReturn(5);
+        ByteBuf data = dummyData();
+        encoder.writeData(ctx, STREAM_ID, data, 10, true, promise);
+        assertEquals(payloadCaptor.getValue().size(), 18);
+        assertTrue(payloadCaptor.getValue().write(18));
+        // writer was called 4 times
+        assertEquals(4, writtenData.size());
+        assertEquals("abcde", writtenData.get(0));
+        assertEquals(0, (int) writtenPadding.get(0));
+        assertEquals("fgh", writtenData.get(1));
+        assertEquals(2, (int) writtenPadding.get(1));
+        assertEquals("", writtenData.get(2));
+        assertEquals(5, (int) writtenPadding.get(2));
+        assertEquals("", writtenData.get(3));
+        assertEquals(3, (int) writtenPadding.get(3));
+        assertEquals(0, data.refCnt());
+    }
+
+    @Test
+    public void emptyFrameShouldSplitPadding() throws Http2Exception {
+        when(frameSizePolicy.maxFrameSize()).thenReturn(5);
+        ByteBuf data = Unpooled.buffer(0);
```

Comment:
I wouldn't expect much different behavior (except for reference count at the end) but should we extract this functionality into a common method and run 2 tests through it.  1 test with `EMPTY_BUFFER`, and 1 test for `Unpooled.buffer(0)`.  Just to make sure we don't have any dependencies on `EMPTY_BUFFER`?